test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ConnectionLimitTests.cs

```diff
@@ -0,0 +1,199 @@
+{
+    public class ConnectionLimitTests
+    {
+        private static async Task KeepOpen(HttpContext context)
+        {
+            // begin the response
+            await context.Response.WriteAsync("Hello");
+
+            // keep it running until aborted
+            while (!context.RequestAborted.IsCancellationRequested)
+            {
+                await Task.Delay(100);
+            }
+        }
+
+        [Fact]
+        public async Task ResetsCountWhenConnectionClosed()
+        {
+            var releasedTcs = new TaskCompletionSource<object>();
+            var lockedTcs = new TaskCompletionSource<bool>();
+            var (context, counter) = SetupMaxConnections(max: 1);
+            counter.OnLock += (s, e) => lockedTcs.TrySetResult(e);
+            counter.OnRelease += (s, e) => releasedTcs.TrySetResult(null);
+
+            using (var server = new TestServer(KeepOpen, context))
+            using (var connection = server.CreateConnection())
+            {
+                await connection.SendEmptyGetAsKeepAlive(); ;
+                await connection.Receive("HTTP/1.1 200 OK");
+                Assert.True(await lockedTcs.Task.TimeoutAfter(TimeSpan.FromSeconds(10)));
+            }
+
+            await releasedTcs.Task.TimeoutAfter(TimeSpan.FromSeconds(10));
+        }
+
+        [Fact]
+        public async Task UpgradedConnectionsCountsAgainstDifferentLimit()
+        {
+            var (testContext, _) = SetupMaxConnections(max: 1);
+            using (var server = new TestServer(async context =>
+            {
+                var feature = context.Features.Get<IHttpUpgradeFeature>();
+                if (feature.IsUpgradableRequest)
+                {
+                    var stream = await feature.UpgradeAsync();
+                    // keep it running until aborted
+                    while (!context.RequestAborted.IsCancellationRequested)
+                    {
+                        await Task.Delay(100);
+                    }
+                }
+            }, testContext))
+            using (var disposables = new DisposableStack<TestConnection>())
+            {
+                var upgraded = server.CreateConnection();
+                disposables.Push(upgraded);
+
+                await upgraded.SendEmptyGetWithUpgrade();
+                await upgraded.Receive("HTTP/1.1 101");
+                // once upgraded, normal connection limit is decreased to allow room for more "normal" connections
+
+                var connection = server.CreateConnection();
+                disposables.Push(connection);
+
+                await connection.SendEmptyGetAsKeepAlive();
+                await connection.Receive("HTTP/1.1 200 OK");
+
+                using (var rejected = server.CreateConnection())
+                {
+                    try
+                    {
+                        // this may throw IOException, depending on how fast Kestrel closes the socket
```

Comment:
Can we make it behave in a consistent way with another tcs? Or is that impossible?